File: js/15686.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');
const [n, m] = inputs[0].split(' ').map(Number);

const map = [];
for (let i = 1; i <= n; i++) {
  map.push(inputs[i].split(' ').map(Number));
}

const homes = [];
const chickenRestaurants = [];

for (let i = 0; i < n; i++) {
  for (let j = 0; j < n; j++) {
    if (map[i][j] === 1) homes.push([i, j]);
    else if (map[i][j] === 2) chickenRestaurants.push([i, j]);
  }
}

const runningRestaurants = [];

let minChickenDistInCity = Infinity;
dfs(0);
console.log(minChickenDistInCity);

function dfs(idx) {
  if (runningRestaurants.length === m) {
    const chickenDistInCity = homes
      .map((home) => {
        const [homeY, homeX] = home;
        const dists = runningRestaurants.map((restaurant) => {
          const [resY, resX] = restaurant;
          return Math.abs(homeY - resY) + Math.abs(homeX - resX);
        });
        return Math.min(...dists);
      })
      .reduce((prev, curr) => prev + curr, 0);
    if (chickenDistInCity < minChickenDistInCity) {
      minChickenDistInCity = chickenDistInCity;
    }
    return;
  }

  for (let i = idx; i < chickenRestaurants.length; i++) {
    runningRestaurants.push(chickenRestaurants[i]);
    dfs(i + 1);
    runningRestaurants.pop();
  }
}

File: c++/15686.cpp
#include <algorithm>
#include <climits>
#include <cmath>
#include <iostream>
#include <vector>

using namespace std;

int n, m;
int map[51][51];
int answer = INT_MAX;
vector<pair<int, int>> homes;
vector<pair<int, int>> chickens;
vector<pair<int, int>> running;

void dfs(int idx) {
  if (running.size() == m) {
    int chikenDistInCity = 0;
    for (int i = 0; i < homes.size(); i++) {
      int homeY = homes[i].first;
      int homeX = homes[i].second;
      int chickenDist = INT_MAX;
      for (int j = 0; j < running.size(); j++) {
        int resY = running[j].first;
        int resX = running[j].second;
        int dist = abs(homeY - resY) + abs(homeX - resX);
        chickenDist = min(chickenDist, dist);
      }
      chikenDistInCity += chickenDist;
    }
    answer = min(answer, chikenDistInCity);
    return;
  }

  for (int i = idx; i < chickens.size(); i++) {
    running.push_back(chickens[i]);
    dfs(i + 1);
    running.pop_back();
  }
}

int main() {
  // freopen("example.txt", "r", stdin);

  cin >> n >> m;

  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      int a;
      cin >> a;
      if (a == 1)
        homes.push_back(make_pair(i, j));
      else if (a == 2)
        chickens.push_back(make_pair(i, j));
    }
  }

  dfs(0);
  cout << answer;
}